### PR TITLE
Incorporate DHooks into base SourceMod

### PR DESCRIPTION
**Pull request by Headline**
Through discussion in `#sourcemod` we are planning on bringing DHooks in-tree, so I guess I volunteered for the dirty work.

There's still concerns about this and I also have yet to scour DHooks source for possible improvements so this is currently WIP.

### CODE REVIEW

**Comment by 404UNFca** [comment]
Next let's add in VoiDeD's transitional helpers and why not a color thing or two while we're at it :smile:

**Comment by KyleSanderson** [comment]
Yikes, style doesn't match core/exts at all :(. I was under the impression this was going to be made more SDKTools like to match the existing API we have before a merge was proposed.

**Comment by Headline** [comment]
@KyleS this is WIP

**Comment by XutaxKamay** [comment]
Hello would it be possible to implement the fix for the native DHookSetReturnVector before completing this pull request?
I've modified it recently for my sourcemod plugin.
https://github.com/XutaxKamay/dhooks/blob/master/vhook.cpp#L744

Previous code was `**(SDKVector**)` when it should be `*(SDKVector*)` I assume

**Comment by qubka** [comment]
Is it possible to create a similar function of SDKCall for call methods with parameters which pass thought registers, Maybe like a generate game data blocks which similar to detour hooks

**Comment by Drifter321** [comment]
This is base dhooks, it does not include detours. So that is outside of the scope of this.

File: extensions/dhooks/vfunc_call.h [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+	unsigned char *vptr = vstk;
+
+	*(void **)vptr = iface;
+
+	if(paramStruct)
+	{
+		vptr += sizeof(void *);
+		paramInfo = (SourceMod::PassInfo *)malloc(sizeof(SourceMod::PassInfo) * dg->params.size());
+
+		for(int i = 0; i < (int)dg->params.size(); i++)
```
Comment (KyleSanderson) [inline extensions/dhooks/vfunc_call.h]: Isn't the real fix here to make this unsigned?

File: extensions/dhooks/vfunc_call.h [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	unsigned char *vstk = (unsigned char *)malloc(sizeof(void *) + size);
+	unsigned char *vptr = vstk;
+
+	*(void **)vptr = iface;
+
+	if(paramStruct)
+	{
+		vptr += sizeof(void *);
+		paramInfo = (SourceMod::PassInfo *)malloc(sizeof(SourceMod::PassInfo) * dg->params.size());
+		for(int i = 0; i < (int)dg->params.size(); i++)
```
Comment (KyleSanderson) [inline extensions/dhooks/vfunc_call.h]: ditto.

**Comment by KyleSanderson** [comment]
@Drifter321 are you okay with re-licensing this project to AM?

File: extensions/dhooks/vfunc_call.h (hunk 1 above)
Comment (Headline) [reply extensions/dhooks/vfunc_call.h]: yes, I intentionally sided on this change as it was the cast that was used in other places throughout DHooks.

My plan going forward is to create another - separate - pull request that will merge changes into this branch that way we can separately review improvements to DHooks as a whole. It'll help with keeping history nice and tidy, unless you suggest that I keep it all in here.

**Comment by Headline** [comment]
> @Drifter321 are you okay with re-licensing this project to AM?

We got approval on Discord from him, but It doesn't hurt to make sure \:)

**Comment by Drifter321** [comment]
👍

File: extensions/dhooks/vhook.cpp [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+			CallVFunction<SDKVector>(dg, paramStruct, g_SHPtr->GetIfacePtr());
+			break;
+		case MRES_Override:
+			if(dg->returnType != ReturnType_Void)
+			{
+				if(returnStruct->isChanged)
+				{
+					g_SHPtr->SetRes(MRES_OVERRIDE);
+					mres = MRES_OVERRIDE;
+					*vec_result = **(SDKVector **)returnStruct->newResult;
```
Comment (Drifter321) [inline extensions/dhooks/vhook.cpp]: Should be *(SDKVector *)

File: extensions/dhooks/vhook.cpp [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+			}
+			break;
+		case MRES_Supercede:
+			if(dg->returnType != ReturnType_Void)
+			{
+				if(returnStruct->isChanged)
+				{
+					g_SHPtr->SetRes(MRES_SUPERCEDE);
+					mres = MRES_SUPERCEDE;
+					*vec_result = **(SDKVector **)returnStruct->newResult;
```
Comment (Drifter321) [inline extensions/dhooks/vhook.cpp]: Should be *(SDKVector *)

File: extensions/dhooks/vhook.cpp [hunk 5]
```diff
@@ -0,0 +1,953 @@
...
+			g_SHPtr->SetRes(MRES_SUPERCEDE);
+			mres = MRES_SUPERCEDE;
+			*vec_result = CallVFunction<SDKVector>(dg, paramStruct, g_SHPtr->GetIfacePtr());
+			break;
+		case MRES_ChangedOverride:
+			if(dg->returnType != ReturnType_Void)
+			{
+				if(returnStruct->isChanged)
+				{
+					*vec_result = **(SDKVector **)returnStruct->newResult;
```
Comment (Drifter321) [inline extensions/dhooks/vhook.cpp]: Should be *(SDKVector *)

**Comment by KyleSanderson** [comment]
@Headline can you do a final sync? We can get Ruben to unblock after that.

**Comment by peace-maker** [comment]
Due to missing SM 1.10 builds of DHooks and a tendency to detour everything instead of using virtual hooks by plugin authors nowadays, I think a lot of servers are running the DHooks version with detouring support. Adding the one without detour natives in tree might cause some conflicts now 🙁

**Comment by KyleSanderson** [comment]
@peace-maker I think the best course of action here based on what's transpired over time is if you open a PR for your fork (as long as you're okay with the relicense to AM for your additions). @Drifter321 has already given his permission for this to be relicensed to AM (thanks man), and we close this one out not as INCOMPLETE but SUPERSEDED.

**Pull request closed without merging**